AppController/lib/taskqueue.rb

```diff
@@ -134,52 +183,59 @@ def self.start_slave(master_ip, clear_data, verbose)
       end
     end
 
-    start_cmds = [
-      # Restarting the RabbitMQ server ensures that we read the correct cookie.
-      "service rabbitmq-server restart",
-      "/usr/sbin/rabbitmqctl stop_app",
-      # Read master hostname given the master IP.
-      "/usr/sbin/rabbitmqctl join_cluster rabbit@#{master_tq_host}",
-      "/usr/sbin/rabbitmqctl start_app"
-    ]
-    full_cmd = "#{start_cmds.join('; ')}"
-    stop_cmd = "/usr/sbin/rabbitmqctl stop"
-    match_cmd = "sname rabbit"
-
+    # We now try to join the master's RabbitMQ cluster. We try in a few
+    # ways: the first one is the simple join, that should work in ideal
+    # condition. The update_cluster_nodes may be needed if the node was
+    # part of the cluster, and it simply needs to rejoin now. The third
+    # one force a reset and a join: this should take care of partitions.
     tries_left = RABBIT_START_RETRY
     loop {
-      MonitInterface.start(:rabbitmq, full_cmd, stop_cmd, [9999], nil,
-                           match_cmd, nil, nil, nil)
-      Djinn.log_debug("Waiting for RabbitMQ on local node to come up")
-      begin
-        Timeout::timeout(MAX_WAIT_FOR_RABBITMQ) do
-          HelperFunctions.sleep_until_port_is_open("localhost", SERVER_PORT)
-          Djinn.log_debug("Done starting rabbitmq_slave on this node")
-
-          Djinn.log_debug("Starting TaskQueue server on slave node")
-          start_taskqueue_server(verbose)
-          Djinn.log_debug("Waiting for TaskQueue server on slave node to come up")
-          HelperFunctions.sleep_until_port_is_open("localhost", 
-                                                   TASKQUEUE_SERVER_INTERNAL_PORT)
-          Djinn.log_debug("Done waiting for TaskQueue server")
-          return
-        end
-      rescue Timeout::Error
-        tries_left -= 1
-        Djinn.log_warn("Waited for RabbitMQ to start, but timed out. " +
-          "Retries left #{tries_left}.")
-        Djinn.log_run("ps ax | grep rabbit | grep -v grep | awk '{print $1}' | xargs kill -9")
-        if clear_data
-          self.erase_local_files()
-        end
+      break if check_taskqueue_server(master_tq_host)
+
+      if join_cluster(:join, master_tq_host)
+        break if check_taskqueue_server(master_tq_host)
+      end
+      if join_cluster(:update, master_tq_host)
+        break if check_taskqueue_server(master_tq_host)
       end
-      if tries_left.zero?
-        Djinn.log_fatal("CRITICAL ERROR: RabbitMQ slave failed to come up")
+      if join_cluster(:reset_join, master_tq_host)
+        break if check_taskqueue_server(master_tq_host)
+      end
+
+      # If we couldn't join the master's cluster, let's back-off a bit and
+      # retry for a little while.
+      tries_left -= 1
+      if tries_left > 0
+        Djinn.log_warn("Retrying to join #{master_tq_host} rabbitmq cluster.")
+        Kernel.sleep(RABBIT_WAIT)
+      else
+        Djinn.log_fatal("CRITICAL ERROR: RabbitMQ slave failed to come up.")
```

Comment:
I believe this log should read something like: "[...] failed to join the cluster."
It's a little misleading as is.